File: Backtracking/Start_Link/C_Start_Link/Start_Link.c
#include <stdio.h>
#include <limits.h>
#include <math.h>

#define MAX 20
#define min(a, b) (((a) < (b)) ? (a) : (b))

int n;
int s[MAX][MAX];
int team[MAX];
int min_diff = INT_MAX;

void get_min_diff(int cnt, int cur)
{
	if (cnt == n / 2)
	{
		int start_tot = 0, link_tot = 0;

		for (int i = 0; i < n; i++)
		{
			for (int j = i + 1; j < n; j++)
			{
				if (team[i] && team[j])
					start_tot += s[i][j] + s[j][i];
				else if (!team[i] && !team[j])
					link_tot += s[i][j] + s[j][i];
			}
		}

		min_diff = min(min_diff, abs(start_tot - link_tot));

		return;
	}

	for (int i = cur; i < n; i++)
	{
		if (!team[i])
		{
			team[i] = 1;
			get_min_diff(cnt + 1, i);
			team[i] = 0;
		}
	}
}

int main()
{
	int team[MAX];

	scanf_s("%d", &n);

	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			scanf_s("%d", &s[i][j]);

	get_min_diff(0, 1);

	printf("%d\n", min_diff);

	return 0;
}

File: Backtracking/Start_Link/Go_Start_link/Start_Link.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

const (
	max = 20
)

var (
	n       = 0
	s       [max][max]int
	team    [max]bool
	minDiff = math.MaxInt64
)

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

func abs(a int) int {
	if a >= 0 {
		return a
	}
	return -a
}

func getMinDiff(cnt, cur int) {
	if cnt == n/2 {
		startTot := 0
		linkTot := 0

		for i := 0; i < n; i++ {
			for j := i + 1; j < n; j++ {
				if team[i] && team[j] {
					startTot += s[i][j] + s[j][i]
				} else if !team[i] && !team[j] {
					linkTot += s[i][j] + s[j][i]
				}
			}
		}

		minDiff = min(minDiff, abs(startTot-linkTot))

		return
	}

	for i := cur; i < n; i++ {
		if !team[i] {
			team[i] = true
			getMinDiff(cnt+1, i)
			team[i] = false
		}
	}
}

func main() {
	br := bufio.NewReader(os.Stdin)
	bw := bufio.NewWriter(os.Stdout)

	fmt.Fscanln(br, &n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			fmt.Fscan(br, &s[i][j])
		}
	}

	getMinDiff(0, 1)

	fmt.Fprintln(bw, minDiff)
	bw.Flush()
}
